src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py

```diff
@@ -739,7 +737,17 @@ def _create(resource_group_name, deployment_name, dns_name_prefix, name, ssh_key
                 "value": client_secret
             }
         }
-    return _invoke_deployment(resource_group_name, deployment_name, template, params, validate, no_wait)
+
+    # Due to SPN replication latency, we do a few retries here
+    maxRetry = 30
+    for i in range(0, maxRetry):
+        try:
+            return _invoke_deployment(resource_group_name, deployment_name, template, params, validate, no_wait)
+        except CloudError as ex:
+            if i != maxRetry - 1 and 'is not valid according to the validation procedure' in ex.message:
+                time.sleep(3)
+            else:
+                raise ex
```

Comment:
You should use the `for-else` structure in this case to avoid the test loop count.

http://python-notes.curiousefficiency.org/en/latest/python_concepts/break_else.html#a-different-kind-of-else

Reply:
Can you be more specific?

how do i `raise ex` in the for-else's else clause?

Reply:
I wrote this example for ya. It combines `for-else` loop and `reuse retry method`.
```
  1 def foo(i):
  2     if i < 5:
  3         raise Exception('from boo')
  4     else:
  5         return 'foo'
  6
  7 def bar(fn, retry_count):
  8     retry_exception = None
  9     for i in range(retry_count):
 10         print('Loop {}'.format(i))
 11         try:
 12             return fn(i)
 13         except Exception as error:
 14             retry_exception = error
 15             continue
 16     else:
 17         raise retry_exception
 18
 19 print('example #1: retry 6 times.')
 20 result = bar(foo, 6)
 21 print('result = {}'.format(result))
 22
 23 print('example #2: retry 3 times. exception is expected.')
 24 result = bar(foo, 3)
~
```

The local index `i` is for demo only. It can be saved.

Reply:
```
   # Due to SPN replication latency, we do a few retries here
    maxRetry = 30
    retry_exception = None
    for _ in range(0, maxRetry):
        try:
            return client.create_or_update(
                resource_group_name=resource_group_name, resource_name=name, parameters=mc, raw=no_wait)
        except CloudError as ex:
            retry_exception = ex
            if 'The credentials in ServicePrincipalProfile were invalid' in ex.message: 
                time.sleep(3)
            else:
                raise ex
    else:
        raise retry_exception

```

Reply:
Turn out, i just don't need the `else`. As there is no place in the for loop will break. So it will be no difference.